components/hub/hub_service.bal

```diff
@@ -0,0 +1,244 @@
+    # + return - `error` if there is any unexpected error or else `()`
+    isolated remote function onSubscriptionIntentVerified(websubhub:VerifiedSubscription message) returns error? {
+        lock {
+            if !message.hasKey(CONSUMER_GROUP) {
+                string consumerGroup = common:generateGroupName(message.hubTopic, message.hubCallback);
+                message[CONSUMER_GROUP] = consumerGroup;
+            }
+            message[SERVER_ID] = config:serverIdentifier;
+            error? persistingResult = persist:addSubscription(message.cloneReadOnly());
+            if persistingResult is error {
+                log:printError("Error occurred while persisting the subscription ", err = persistingResult.message());
+            }
+        }
+    }
+
+    # Unsubscribes a `subscriber` from the hub.
+    #
+    # + message - Details of the unsubscription
+    # + headers - `http:Headers` of the original `http:Request`
+    # + return - `websubhub:UnsubscriptionAccepted` if unsubscription is accepted from the hub, `websubhub:BadUnsubscriptionError`
+    # if unsubscription is denied from the hub or `error` if there is any unexpected error
+    isolated remote function onUnsubscription(websubhub:Unsubscription message, http:Headers headers)
+                returns websubhub:UnsubscriptionAccepted|websubhub:BadUnsubscriptionError|error {
+        if config:securityOn {
+            check security:authorize(headers, ["subscribe"]);
+        }
+        return websubhub:UNSUBSCRIPTION_ACCEPTED;
+    }
+
+    # Validates a incomming unsubscription request.
+    #
+    # + message - Details of the unsubscription
+    # + return - `websubhub:UnsubscriptionDeniedError` if the unsubscription is denied by the hub or else `()`
+    isolated remote function onUnsubscriptionValidation(websubhub:Unsubscription message)
+                returns websubhub:UnsubscriptionDeniedError? {
+        boolean topicAvailable = false;
+        lock {
+            topicAvailable = registeredTopicsCache.hasKey(message.hubTopic);
+        }
+        if !topicAvailable {
+            return error websubhub:UnsubscriptionDeniedError(
+                "Topic [" + message.hubTopic + "] is not registered with the Hub", statusCode = http:STATUS_NOT_ACCEPTABLE);
+        } else {
+            string subscriberId = common:generateSubscriberId(message.hubTopic, message.hubCallback);
+            if !isValidSubscription(subscriberId) {
+                return error websubhub:UnsubscriptionDeniedError("Could not find a valid subscriber for Topic ["
+                                + message.hubTopic + "] and Callback [" + message.hubCallback + "]", statusCode = http:STATUS_NOT_ACCEPTABLE);
+            }
+        }
+    }
+
+    # Processes a verified unsubscription request.
+    #
+    # + message - Details of the unsubscription
+    isolated remote function onUnsubscriptionIntentVerified(websubhub:VerifiedUnsubscription message) {
+        lock {
```

Comment:
Why do we need `lock` here?

Reply:
Fixed with https://github.com/wso2/product-integrator-websubhub/pull/1/commits/84204db7e1afc92107250781335a0bf81540750b